robosuite/controllers/base_controller.py

```diff
@@ -0,0 +1,98 @@
+import abc
+import numpy as np
+import mujoco_py
+
+
+class Controller(object, metaclass=abc.ABCMeta):
+    """
+    General controller interface.
+
+    Requires reference to mujoco sim object, id_name of specific robot, relevant joint_indexes to that robot, and
+    whether an initial_joint is used for nullspace torques or not
+    """
+    def __init__(self,
+                 sim,
+                 id_name,
+                 joint_indexes,
+                 initial_joint
+    ):
+        self.action_scale = None
+
+        # Note: the following attributes must be specified in extended controller subclasses during the init call:
+        # self.output_min
+        # self.output_max
+        # self.input_min
+        # self.input_max
+
+        # mujoco simulator state
+        self.sim = sim
+        self.id_name = id_name
+        self.joint_index = joint_indexes
+
+        # robot states
+        self.ee_pos = None
+        self.ee_ori_mat = None
+        self.ee_pos_vel = None
+        self.ee_ori_vel = None
+        self.joint_pos = None
+        self.joint_vel = None
+
+        # dynamics and kinematics
+        self.J_pos = None
+        self.J_ori = None
+        self.J_full = None
+        self.mass_matrix = None
+
+        # Joint dimension
+        self.joint_dim = len(joint_indexes)
+
+        self.update()
+        self.initial_joint = initial_joint
+        print("initial_joint: {}".format(self.initial_joint))       # TODO: Clean post-debugging
+        self.torque_compensation = np.zeros(self.joint_dim)
+
+    @abc.abstractmethod
+    def run_controller(self, action):
+        """
+        Go from actions to torques
+        """
+        pass
+
+    def scale_action(self, action):
+        """
+        Scale the action based on max and min of action
+        """
+
+        if self.action_scale is None:
+            self.action_scale = abs(self.output_max - self.output_min) / abs(self.input_max - self.input_min)
+            self.action_output_transform = (self.output_max + self.output_min) / 2.0
+            self.action_input_transform = (self.input_max + self.input_min) / 2.0
+        action = np.clip(action, self.input_min, self.input_max)
+        transformed_action = (action - self.action_input_transform) * self.action_scale + self.action_output_transform
+
+        return transformed_action
+
+    def update(self):
+
+        self.model_timestep = self.sim.model.opt.timestep
+
+        self.ee_pos = self.sim.data.body_xpos[self.sim.model.body_name2id(self.id_name)]
```

Comment:
for all of these where you are reading from self.sim, it might be safer to wrap these in a np.array call to ensure copies are made